File: ios/CareLog/CareLog/Notifications/NotificationDeepLinkHandler.swift
import Foundation

/// Handles deep linking from push notifications.
struct NotificationDeepLinkHandler {

    /// Deep link data extracted from notification.
    struct DeepLinkData {
        let type: NotificationType
        let patientId: String?
        let vitalType: String?
        let alertId: String?
    }

    /// Types of notifications.
    enum NotificationType: String {
        case thresholdBreach = "THRESHOLD_BREACH"
        case reminderLapse = "REMINDER_LAPSE"
        case patientReminder = "PATIENT_REMINDER"
        case system = "SYSTEM"
    }

    /// Navigation destinations for deep linking.
    enum NavigationDestination {
        case dashboard
        case alertInbox
        case bloodPressureLog
        case glucoseLog
        case temperatureLog
        case weightLog
        case pulseLog
        case spo2Log
    }

    /// Parse deep link data from notification user info.
    static func parseDeepLink(from userInfo: [AnyHashable: Any]) -> DeepLinkData? {
        // Try to get data from APNs payload
        let data: [String: Any]
        if let apsData = userInfo["data"] as? [String: Any] {
            data = apsData
        } else {
            data = userInfo as? [String: Any] ?? [:]
        }

        guard let typeString = data["type"] as? String,
              let type = NotificationType(rawValue: typeString) else {
            return nil
        }

        return DeepLinkData(
            type: type,
            patientId: data["patientId"] as? String,
            vitalType: data["vitalType"] as? String,
            alertId: data["alertId"] as? String
        )
    }

    /// Determine the navigation destination based on deep link type.
    static func getNavigationDestination(for deepLink: DeepLinkData) -> NavigationDestination {
        switch deepLink.type {
        case .thresholdBreach, .reminderLapse:
            return .alertInbox

        case .patientReminder:
            // Navigate to the appropriate vital logging screen
            switch deepLink.vitalType {
            case "BLOOD_PRESSURE":
                return .bloodPressureLog
            case "GLUCOSE":
                return .glucoseLog
            case "TEMPERATURE":
                return .temperatureLog
            case "WEIGHT":
                return .weightLog
            case "PULSE":
                return .pulseLog
            case "SPO2":
                return .spo2Log
            default:
                return .dashboard
            }

        case .system:
            return .dashboard
        }
    }
}

File: android/app/src/main/java/com/carelog/notifications/NotificationDeepLinkHandler.kt
package com.carelog.notifications

import android.content.Intent
import android.os.Bundle

/**
 * Handles deep linking from push notifications.
 */
object NotificationDeepLinkHandler {

    /**
     * Notification deep link data.
     */
    data class DeepLinkData(
        val type: String?,
        val patientId: String?,
        val vitalType: String?,
        val alertId: String?
    )

    /**
     * Extract deep link data from intent extras.
     */
    fun extractDeepLinkData(intent: Intent?): DeepLinkData? {
        val extras = intent?.extras ?: return null

        val type = extras.getString("type")
        if (type == null) return null

        return DeepLinkData(
            type = type,
            patientId = extras.getString("patientId"),
            vitalType = extras.getString("vitalType"),
            alertId = extras.getString("alertId")
        )
    }

    /**
     * Extract deep link data from bundle.
     */
    fun extractDeepLinkData(bundle: Bundle?): DeepLinkData? {
        if (bundle == null) return null

        val type = bundle.getString("type")
        if (type == null) return null

        return DeepLinkData(
            type = type,
            patientId = bundle.getString("patientId"),
            vitalType = bundle.getString("vitalType"),
            alertId = bundle.getString("alertId")
        )
    }

    /**
     * Determine the navigation destination based on deep link type.
     */
    fun getNavigationDestination(deepLinkData: DeepLinkData): NavigationDestination {
        return when (deepLinkData.type) {
            "THRESHOLD_BREACH" -> NavigationDestination.AlertInbox
            "REMINDER_LAPSE" -> NavigationDestination.AlertInbox
            "PATIENT_REMINDER" -> {
                // Navigate to the appropriate vital logging screen
                when (deepLinkData.vitalType) {
                    "BLOOD_PRESSURE" -> NavigationDestination.BloodPressureLog
                    "GLUCOSE" -> NavigationDestination.GlucoseLog
                    "TEMPERATURE" -> NavigationDestination.TemperatureLog
                    "WEIGHT" -> NavigationDestination.WeightLog
                    "PULSE" -> NavigationDestination.PulseLog
                    "SPO2" -> NavigationDestination.SpO2Log
                    else -> NavigationDestination.Dashboard
                }
            }
            else -> NavigationDestination.Dashboard
        }
    }

    /**
     * Navigation destinations for deep linking.
     */
    sealed class NavigationDestination {
        data object Dashboard : NavigationDestination()
        data object AlertInbox : NavigationDestination()
        data object BloodPressureLog : NavigationDestination()
        data object GlucoseLog : NavigationDestination()
        data object TemperatureLog : NavigationDestination()
        data object WeightLog : NavigationDestination()
        data object PulseLog : NavigationDestination()
        data object SpO2Log : NavigationDestination()
    }
}
